### PR TITLE
Handle multiple commands per line; move toward a parser

### PR DESCRIPTION
**Pull request by skylerberg**
Details in commit messages.

### CODE REVIEW

**Comment by jackpot51** [comment]
Make sure you have lots of tests! Shellshock was added to bash at around this stage

**Comment by Ticki** [comment]
We actually need better testing a lot of places.

**Comment by skylerberg** [comment]
I have added quite a few tests so far. I am trying to break the large untested main file into modular parts that are each tested.

**Comment by jackpot51** [comment]
The tests here are very good so far, thanks

File: src/expansion.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+use std::collections::BTreeMap;
+use super::tokenizer::Token;
+
+pub fn expand_tokens(tokens: &mut Vec<Token>, variables: &BTreeMap<String, String>) -> Vec<Token> {
```
Comment (hauleth) [inline src/expansion.rs]: You should use `&mut [T]` instead of `&mut Vec<T>`.

**Comment by hauleth** [comment]
Why use custom parser instead of existing ones like `nom`?

**Comment by skylerberg** [comment]
I hadn't heard of `nom`. It looks neat! I will take a closer look and hopefully we can use it.

**Comment by Ticki** [comment]
Isnt nom overkill for this task?

**Comment by hauleth** [comment]
We are creating language here. Why writing your own buggy code when someone can do this for you?

**Comment by Ticki** [comment]
Yeah, you're right. Nom is fine for this task.

**Comment by hauleth** [comment]
There are also other solutions like [Lalrpop](https://github.com/nikomatsakis/lalrpop), but I find Nom simplest and easiest.

**Comment by Ticki** [comment]
@hauleth Lalrpop is LR(1), but nom is not. I doubt Lalrpop will be good for this task (if we want a non-trivial language).

**Pull request merged**